### PR TITLE
feat: Rewrite logical expression to match InfluxQL behaviour

### LINKED ISSUE #6895: InfluxQL-IOx: Type cast post fix operator
Teach the InfluxQL planner how to generate a LogicalPlan that casts between numeric types by transforming the InfluxQL postfix typecast operator.

# Background

InfluxQL `SELECT` statements support the ability to cast numeric expressions to other numeric types. For example, the following statement will explicitly cast the float field, `float_field`, to a 64-bit integer, and the integer field, `integer_field`, to a 64-bit float.

```sql
SELECT float_field::integer, integer_field::float FROM measurement
```

Attempting to cast tags, or non-numeric fields results in a `NULL` value.

### LINKED ISSUE #6912: InfluxQL-IOx: Teach the planner how to rewrite expression trees
Teach the InfluxQL planner how to rewrite an expression tree such that it respects the following properties of the InfluxQL query engine:

* Invalid column references evaluate to to the literal value `NULL`
* Incompatible types are always coerced to a compatible value, `NULL` or `false`, depending on the operation.

Regular expression conditional expressions are exempt from these rules, as they only permit the simple form of `<tag or field> regex_op /regex/`

> **NOTE**  
> The planner is capable of reducing time-range expressions:
>
> https://github.com/influxdata/influxdb_iox/blob/bd2a72a4b65e900b9f8d2277744facec7418f97c/iox_query/src/plan/influxql/planner_time_range_expression.rs#L86

therefore, there may be an opportunity to reuse this code.

# Background

InfluxQL tries hard to execute a query without returning an error. InfluxQL attempts to coerce differing data types to compatible types to perform binary operations, or evaluates expressions to `false` or `nil` if this is not possible. InfluxQL also evaluates expressions referring to a missing `VarRef` (tag or field) as false.

Filtering is first evaluated using the time-series index for any conditional expressions involving tag keys:

https://github.com/influxdata/influxdb/blob/82d1123e787d9562c14f872d234d43a6641b3478/tsdb/engine/tsm1/engine.go#L2532

This process efficiently eliminates series keys from entering the InfluxQL query engine.

However, this the time-series index is limited to simple expressions, of the form `tag OP literal` along with the conjunction (`AND`) and disjunction (`OR`) operators. For complex expressions, evaluating and filtering is handed off to the InfluxQL query engine. 

The query engine, using the filtered set of series keys from TSI, eagerly constructs a TSM _iterator_ for every series key. An iterator represents a projection in the `SELECT` statement. A series key is a combination of the tag set and a single field and reads the time-series data (timestamps and values) from TSM files. We can observe the implementation of the query engine filtering of a specific iterator type, in this case float64:

https://github.com/influxdata/influxdb/blob/dee8977d2c6598cb2d17e9334ea997c99853640a/tsdb/engine/tsm1/iterator.gen.go#L280-L283

Which calls the `ValuerEval.EvalBool` function:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4181-L4183

Most of the implementation details for evaluating binary expressions are found in the `evalBinaryExpr` function:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4188

Lets examine evaluating the expression `float64 > string`. We reach this case in the switch statement, which determines the LHS is a `float64`:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4224

next, InfluxQL attempts to coerce the RHS to a `float64`, where `ok` will be `false` in this case as the RHS data type is `string`:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4226-L4234

Finally, we switch on `OP` which is `Gt` and short circuit on `ok` as it is false, returning `false`:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4246-L4247

Other operators, such as `ADD` result in a `nil` value:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4250-L4253

> **NOTE**  
> The `ValuerEval` includes a reference to a `Valuer`:
>
> https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4130

which, among others, is responsible for mapping keys to values. In this case, the keys are a `VarRef` symbol (a field or tag), mapped to their value:

https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4173-L4174

The InfluxQL query iterators, such as the `floatIterator` uses a composite `Valuer`:

https://github.com/influxdata/influxdb/blob/dee8977d2c6598cb2d17e9334ea997c99853640a/tsdb/engine/tsm1/iterator.gen.go#L228-L231

Which allows it to also evaluate math functions. The `MapValuer` references a map, which is updated with only additional auxiliary field values[^aux_fields] on each iteration of the iterator:

https://github.com/influxdata/influxdb/blob/dee8977d2c6598cb2d17e9334ea997c99853640a/tsdb/engine/tsm1/iterator.gen.go#L277

Given the `MapValuer` _does not include tag keys and values_. This means that the InfluxQL query engine cannot evaluate complex expressions such as `tag0 + tag1 = 'string'`, which always evaluate to `false`.

[^aux_fields]: These are additional TSM iterators for other fields of the same series key, joined on timestamp to support evaluating predicates on fields in a `WHERE` clause

### PR DESCRIPTION
**Pull request by stuartcarnie**
## Rewrite logical expressions

Closes #6912 

The primary goal of this PR is to teach the InfluxQL logical planner how to rewrite logical expressions such that they match InfluxQL behaviour. This InfluxQL compatibility treatment is performed on conditional expressions and projection expressions. The bulk of this logic is found in the `planner_rewrite_expression.rs` module and is accompanied by a number of tests to verify the expected behaviour.

## Learns cast postfix operator

Closes #6895

The InfluxQL planner also learned how to interpret the cast postfix operator, which was mostly completed as part of #6912. The most significant improvement was learning about the `unsigned` data type.

## 🛠️ Maintenance

There are a few other maintenance items included with this PR:

* Whilst working on #7011, I learned about a number of DataFusion helper functions to reduce the tedium of constructing logical expressions, which I have utilised. For example the following

  https://github.com/influxdata/influxdb_iox/blob/b1b7865d351e159dd09f431c85bd9eedb3e37385/iox_query/src/plan/influxql/planner.rs#L321-L325

  is simplified to

  https://github.com/influxdata/influxdb_iox/blob/d33b619fb17d7ac55bf13cb34bd24f60c9e2068a/iox_query/src/plan/influxql/planner.rs#L339

  There are a number of other cases that were improved.

* Moved the `insta` snapshots to inline tests, which is arguably an improved development experience, given the expected output appears inline with the test code.

### CODE REVIEW

File: iox_query/src/plan/influxql/planner.rs [hunk 1]
```diff
@@ -923,11 +944,154 @@ mod test {
...
+                TableScan: all_types [bool_field:Boolean;N, f64_field:Float64;N, i64_field:Int64;N, str_field:Utf8;N, tag0:Dictionary(Int32, Utf8);N, tag1:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None), u64_field:UInt64;N]
+            "###);
+            assert_snapshot!(plan("SELECT f64_field::string FROM all_types").await, @r###"
+            Projection: all_types.time, NULL AS f64_field [time:Timestamp(Nanosecond, None), f64_field:Null;N]
+              Sort: all_types.time ASC NULLS LAST [bool_field:Boolean;N, f64_field:Float64;N, i64_field:Int64;N, str_field:Utf8;N, tag0:Dictionary(Int32, Utf8);N, tag1:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None), u64_field:UInt64;N]
+                TableScan: all_types [bool_field:Boolean;N, f64_field:Float64;N, i64_field:Int64;N, str_field:Utf8;N, tag0:Dictionary(Int32, Utf8);N, tag1:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None), u64_field:UInt64;N]
+            "###);
+            assert_snapshot!(plan("SELECT f64_field::boolean FROM all_types").await, @r###"
```
Comment (alamb) [inline iox_query/src/plan/influxql/planner.rs:969]: That is interesting that fasting from f64 to boolean is allowed (though it appears to result in NULL, which makes sense)

I thinK DataFusion would straight up reject this type of query with an error

**Review by alamb — APPROVED** [review]
I went through this PR carefully and it makes sense to me 

> The primary goal of this PR is to teach the InfluxQL logical planner how to rewrite logical expressions such that they match InfluxQL behaviour. This InfluxQL compatibility treatment is performed on conditional expressions and projection expressions. The bulk of this logic is found in the planner_rewrite_expression.rs module and is accompanied by a number of tests to verify the expected behaviour.

I didn't see any changes to a `planner_rewrite_expression` module, and I didn't see a lot of logic to rewrite expressions -- instead I saw better support and tests for casting as well as new support for unsigned.

File: iox_query/src/plan/influxql/planner_rewrite_expression.rs [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+//!
+//! ```text
+//! > select uptime_format from system where uptime_format + '' = '5 days,  ' + '7:08';
+//! <null>
+//! ```
+//!
+//!
+//! [`Reduce`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4850-L4852
+//! [`EvalBool`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4181-L4183
+//! [`Eval`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4137
```
Comment (wolffcm) [inline iox_query/src/plan/influxql/planner_rewrite_expression.rs:123]: It never ceases to surprise me how much unintuitive behavior there is in InfluxQL. I really appreciate the work you've done to characterize it.

**Review by wolffcm — APPROVED** [review]
Looks good!

File: iox_query/src/plan/influxql/planner.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner.rs:969]: I would have thought so too, but TIL:

```
❯ select 5.4::boolean, 0.0::boolean;
+--------------+------------+
| Float64(5.4) | Float64(0) |
+--------------+------------+
| true         | false      |
+--------------+------------+
1 row in set. Query took 0.001 seconds.
```

File: iox_query/src/plan/influxql/planner_rewrite_expression.rs (hunk 2 above)
Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner_rewrite_expression.rs:123]: No worries – it has been quite the learning experience for me too. From past usages of InfluxQL, I've recalled some frustrations and now I have answers as to why it works the way it does 😊 

I look forward to the opportunity to change those cases to more intuitive behaviour, and some I already have, such as being able to use `time ` in `OR` conditions, such as `time = <val> OR time = <val>`, which isn't supported in InfluxQL today. 

There is an exciting opportunity to move InfluxQL forward and leverage the power and features available to us in DataFusion.

Comment (alamb) [reply iox_query/src/plan/influxql/planner_rewrite_expression.rs:123]: Yeah, for all its quirks, there are certain timeseries operations where InfluxQL is far superior at expressing. I can't wait to see it supported natively in IOx

**Pull request merged**